### PR TITLE
feat: add `evolve` function 

### PR DESCRIPTION
**Pull request by fneco**
close https://github.com/remeda/remeda/issues/452

---

Make sure that you:

- [x] Typedoc added for new methods and updated for changed
- [x] Tests added for new methods and updated for changed
- [x] New methods added to `src/index.ts`
- [x] New methods added to `mapping.md`

---

<details><summary>We use semantic PR titles to automate the release process!</summary>

https://conventionalcommits.org

PRs should be titled following using the format: `< TYPE >(< scope >)?: description`

### Available Types:

- `feat`: new functions, and changes to a function's type that would impact users.
- `fix`: changes to the runtime behavior of an existing function, or refinements to it's type that shouldn't impact most users.
- `perf`: changes to function implementations that improve a functions _runtime_ performance.
- `refactor`: changes to function implementations that are neither `fix` nor `perf`
- `test`: tests-only changes (transparent to users of the function).
- `docs`: changes to the documentation of a function **or the documentation site**.
- `build`, `ci`, `style`, `chore`, and `revert`: are only relevant for the internals of the library.

For scope put the name of the function you are working on (either new or
existing).

</details>

### CODE REVIEW

File: src/evolve.test.ts [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+    const expected = { n: 0 };
+    const result = evolve(object, transf);
+    expect(result).toEqual(expected);
+    expectTypeOf(result).toEqualTypeOf<typeof expected>();
+  });
+
+  it('creates a new array by evolving the `array` according to the `transformation` functions', function () {
+    // NOTE:
+    // If we use tuple in `transformations` parameter,
+    // use `as const` or `evolve` can't handle typing.
```
Comment (fneco) [inline src/evolve.test.ts:77]: We could stop handling functions in arrays, but for now, we are adapting to the Ramda implementation.

File: src/evolve.test.ts [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+    const expected = [1, 2, 3, ['...', 'Go!!']] as const;
+    let result = evolve(object, transf);
+    expect(result).toEqual(expected);
+    // FIXME: Because of limitations of working with tuple types in TypeScript.
+    expectTypeOf(result).toEqualTypeOf<{
+      0: number;
+      1: 2;
+      2: number;
+      3: { 0: '...'; 1: string };
+    }>();
```
Comment (fneco) [inline src/evolve.test.ts:96]: It might be possible to FIX it by applying `ObjectToTuple` in the [comment here](https://stackoverflow.com/a/73290716), but it would be more complicated.

Should I stop dealing with functions in arrays or keep the original Ramda implementation?

**Comment by eranhirsch** [comment]
Hey, just wanted to ping that I didn't get to reviewing this yet and I'm sorry for the delays. I will try to take a look at it now, but might not finish the review...

**Comment by fneco** [comment]
I'm happy to hear from you. Thank you for your concern. I will wait patiently.

**Review by eranhirsch — CHANGES_REQUESTED** [review]
The typing of the evolve function you implemented here doesn't provide enough type-safety and ergonomics that this method could offer. The issue stems from the fact that you build "backwards" from the transformations object to the evolved object, as can be seen in your test examples, eg:

```ts
const transf = {
  count: add(1),
  data: { elapsed: add(1), remaining: add(-1) },
};
const object = {
  id: 10,
  count: 10,
  data: { elapsed: 100, remaining: 1400 },
};
```

The transform object here is not typed in relation to the object, each one of them is typed on their own.

Instead, typing the transformation object **from** the main data object would allow typescript to infer the types for the transformation object so they are tightly coupled.

here's an example that fails with your typing:
```ts
  test('myStuff', () => {
    const data: { animalType: 'cat' | 'dog' } = { animalType: 'cat' };
    evolve(data, {
      animalType: type => {
        console.log(type);
        return type;
      },
    });
  });
```

In short, take the typing of the input, T, and construct the type for the transformations object from it, and not the other way around.

**Comment by fneco** [comment]
Thank you for your kind review.

## ❶ about type

> In short, take the typing of the input, T, and construct the type for the transformations object from it, and not the other way around.

I change the implementation like this:
* In case of Data First, data arg type construct the transformations object type (I added new `Evolver` type)
* In case of Data Last, vice versa, transformations object type construct data arg type (in the same manner as before)

As a result, the noted problems were resolved as shown in the image below.

image

## ❷ about changing implementation

As I mentioned [in past comments](https://github.com/remeda/remeda/pull/454#discussion_r1446832361),
I found it difficult to meet Remeda's requirements for type safety while maintaining the lambda implementation.
So, I remove functionality handling functions tuple (e.g. `[add(1), add(-1) ]`).
Accordingly, the type and implementation were modified.


We apologize for the inconvenience, but we would appreciate a re-review.

File: src/evolve.ts [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+  let transformation, key, type;
+  for (key in object) {
+    transformation = transformations[key];
+    type = typeof transformation;
+    result[key] =
+      type === 'function'
+        ? transformation(object[key])
+        : transformation && type === 'object'
+          ? _evolve(object[key], transformation)
+          : object[key];
```
[the remark this answers is not in the document]
Comment (fneco) [reply src/evolve.ts]: Thank you for your advice, I integrated your suggestions into the source code.

File: src/evolve.ts [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+export function evolve<E>(
+  transformations: E
+): <T extends EvolveTargetObject<E>>(object: T) => Evolve<T, E>;
+
+export function evolve() {
+  return purry(_evolve, arguments);
+}
+
+function _evolve(object: any, transformations: any) {
+  if (!isObject(object) && !isArray(object)) {
```
[the remark this answers is not in the document]
Comment (fneco) [reply src/evolve.ts]: Thanks for letting me know.

File: src/evolve.ts [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    return object;
+  }
+  const result: Record<string, any> = object instanceof Array ? [] : {};
+  let transformation, key, type;
+  for (key in object) {
+    transformation = transformations[key];
```
[the remark this answers is not in the document]
Comment (fneco) [reply src/evolve.ts]: I have re-implemented it following that approach.

File: src/evolve.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (fneco) [reply src/evolve.ts]: I Understood.

File: src/evolve.ts [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+ */
+export function evolve<E>(
+  transformations: E
+): <T extends EvolveTargetObject<E>>(object: T) => Evolve<T, E>;
+
+export function evolve() {
+  return purry(_evolve, arguments);
+}
+
+function _evolve(object: any, transformations: any) {
```
[the remark this answers is not in the document]
Comment (fneco) [reply src/evolve.ts]: I implemented it using the approach of using types as much as possible, and the result is shown [here](https://github.com/remeda/remeda/blob/541bb47603431805696ee7844e7089af7da481e7/src/evolve.ts#L248~L280).
Notations such as `result[key]` are strongly restricted by the result type and key type, so to avoid this, I separated array processing and object processing.
However, in the end, as described in [the comments here ❷](https://github.com/remeda/remeda/pull/454#issuecomment-1940654791), I decided not to handle arrays.

[the remark this answers is not in the document]
Comment (fneco) [reply src/evolve.ts]: Exactly as you said. I have made the correction.

File: src/evolve.ts [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+  if (!isObject(object) && !isArray(object)) {
+    return object;
+  }
+  const result: Record<string, any> = object instanceof Array ? [] : {};
```
[the remark this answers is not in the document]
Comment (fneco) [reply src/evolve.ts]: I Understood.

File: src/evolve.ts [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+ */
+export function evolve<E extends EvolverStructure>(
+  evolver: E
+): <T extends EvolveTarget<E>>(object: T) => Evolved<T, E>;
+
+export function evolve() {
+  return purry(_evolve, arguments);
+}
+
+function _evolve(data: NestedObject, evolver: EvolverStructure): NestedObject {
```
Comment (eranhirsch) [inline src/evolve.ts:209]: As mentioned above, NestedObject is overkill for this function, simply use `unknown` in the signature and rely on the test below (`typeof data !== "object" || data === null`) to get an `object` which you can spread.

```suggestion
function _evolve(data: unknown, evolver: GenericEvolver): unknown{
```

File: src/evolve.ts [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+  if (typeof data !== 'object' || data === null) {
+    return data; // Dead logic if the type is followed.
+  }
+
+  const dataKeys = Object.keys(data);
+
+  return toPairs.strict(evolver).reduce<NestedObject>(
```
Comment (eranhirsch) [inline src/evolve.ts:216]: Just use `Record<string, unknown>` and you'll be able to do everything you need without typescript issues
```suggestion
  return toPairs.strict(evolver).reduce<Record<string, unknown>>(
```

File: src/evolve.ts [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+  ? GetFirstParam<E>
+  : E extends object
+    ? {
+        [K in keyof E as E[K] extends AFunction
+          ? K
+          : E[K] extends object
+            ? K
+            : never]?: EvolveTarget<E[K]>;
+      }
+    : never;
```
Comment (eranhirsch) [inline src/evolve.ts:77]: You don't need any of this, the whole function could be implemented with just Evolver and Evolved.

```suggestion
```

File: src/evolve.ts [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+ */
+type Evolver<T> = [T] extends [Array<any>]
+  ? (data: T) => any
+  : T extends object
+    ? {
+        [K in keyof T]?:
+          | ([T[K]] extends [Array<any>] ? never : (data: T[K]) => any)
+          | Evolver<T[K]>;
+      }
+    : never;
```
Comment (eranhirsch) [inline src/evolve.ts:115]: 1. You are over-complicating things! You don't need all these fancy tests (`[T] extends [...]`). These types show up in type-fest because that library is aimed at providing excellent types even for really "academic" edge-cases. We don't encounter issues and reports from users running into problems without them.
2. Notice that you need to make T `Required` before taking its value. Otherwise, the evolver functions would take `undefined` for optional props (which is not how the runtime behavior handles them).
3. You don't need to handle the array specifically; it will be folded into the object case, which is how the runtime would see things.

This version of Evolver works with all your tests:
```suggestion
type Evolver<T> = T extends object
  ? {
      readonly [K in keyof T]?:
        | ((data: Required<T>[K]) => unknown)
        | Evolver<T[K]>;
    }
  : never;
```

File: src/evolve.ts [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+/**
+ * Creates return type from the type of arguments of `evolve`.
+ */
+type Evolved<T, E> = E extends AFunction
+  ? ReturnType<E>
+  : T extends object
+    ? {
+        [K in keyof T]: Evolved<T[K], GetValueByKey<E, K, T[K]>>;
+      }
+    : T;
```
Comment (eranhirsch) [inline src/evolve.ts:126]: And similarly this type doesn't need to be this complicated:

```suggestion
type Evolved<T, E> = T extends object
  ? {
      [K in keyof T]: K extends keyof E
        ? E[K] extends (...arg: any) => infer R
          ? R
          : Evolved<T[K], E[K]>
        : T[K];
    }
  : T;
```

File: src/evolve.ts [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+ *    //   id: 10,
+ *    //   count: 11,
+ *    //   time: { elapsed: 101, remaining: 1399 },
+ *    // }
+ * @dataLast
+ * @category Object
+ */
+export function evolve<E extends EvolverStructure>(
+  evolver: E
+): <T extends EvolveTarget<E>>(object: T) => Evolved<T, E>;
```
Comment (eranhirsch) [inline src/evolve.ts:203]: This "backwards" inference is over-complicating things. dataLast usages are mainly for pipes or places where the type of T is already inferred. In those cases you don't need to construct an "EvolveTarget" from the evolver. Simply use the same types you use for dataFirst.
```suggestion
export function evolve<T, E extends Evolver<T>>(
  evolver: E
): (object: T) => Evolved<T, E>;
```

**Review by eranhirsch — CHANGES_REQUESTED** [review]
This is a complex function to get right!

You are over-complicating the task and introducing more when we can solve it with less. I've implemented the changes I propose here locally on my machine and all tests ran successfully, so none of the things I suggested to remove seem necessary.

File: src/evolve.ts [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+ * basic structure of `data` parameter and return of the function `evolve`.
+ */
+type NestedObject = Record<string, Primitive | object>;
+
+/**
+ * basic structure of `evolver` parameter of the function `evolve`.
+ */
+type EvolverStructure = Readonly<
+  Record<string, ((data: unknown) => unknown) | object | undefined>
+>;
```
[the remark this answers is not in the document]
Comment (fneco) [reply src/evolve.ts:17]: it's great!

File: src/evolve.ts [hunk 15]
```diff
@@ -0,0 +1,236 @@
+import { purry } from './purry';
+import { toPairs } from './toPairs';
+
+type AFunction = (...a: Array<any>) => any;
+type Primitive = string | number | boolean | null | undefined | symbol;
+
+/**
+ * basic structure of `data` parameter and return of the function `evolve`.
+ */
+type NestedObject = Record<string, Primitive | object>;
```
[the remark this answers is not in the document]
Comment (fneco) [reply src/evolve.ts:10]: I'm sorry I couldn't handle it better and made it more complicated.

File: src/evolve.ts [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+    return data; // Dead logic if the type is followed.
+  }
+
+  const dataKeys = Object.keys(data);
+
+  return toPairs.strict(evolver).reduce<NestedObject>(
+    (result, [key, value]) => {
+      if (dataKeys.indexOf(key) === -1) {
+        return result;
+      }
```
[the remark this answers is not in the document]
Comment (fneco) [reply src/evolve.ts:220]: Thank you.

File: src/evolve.ts [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+      if (typeof value === 'function') {
+        result[key] = value(result[key]);
+        return result;
+      }
+      if (value === null || value === undefined) {
+        return result;
+      }
```
[the remark this answers is not in the document]
Comment (fneco) [reply src/evolve.ts:227]: Indeed, I would rather have a runtime error when the type is ignored.

[the remark this answers is not in the document]
Comment (fneco) [reply src/evolve.ts:227]: There was an oversight, thank you.

File: src/evolve.ts (hunk 8 above)
Comment (fneco) [reply src/evolve.ts:209]: Thank you very much.

File: src/evolve.ts (hunk 10 above)
Comment (fneco) [reply src/evolve.ts:77]: Indeed, it passed the type test without it.

File: src/evolve.ts (hunk 13 above)
Comment (fneco) [reply src/evolve.ts:203]: I was lacking in knowledge. Thank you for teaching me this way of writing.

File: src/evolve.ts (hunk 11 above)
Comment (fneco) [reply src/evolve.ts:115]: 1. When I was doing my trial and error, this solved a certain problem, but was not necessary with the current code. Thanks for bringing this to my attention.
2. Thank you very much! So this will prevent the following code from causing a type error!
```ts
  it('does not invoke function if `data` does not contain the key', function () {
    const data: { id?: number } = {};
    const expected = {};
    const transf = { id: add(1)};
    const result = evolve(data, transf);
    expect(result).toEqual(expected);
  });
```
3. [please pay attension] As I said [here in ❷](https://github.com/remeda/remeda/pull/454#issuecomment-1940654791), I changed implementation so that `evolver` cannot handle arrays of functions. So, I want to get type error in following pattern.
```ts
  it('can not handle function arrays.', function () {
    const expected = {
      quartile: [2, 1],
    };
    const result = evolve(
      {
        quartile: [1, 2],
      },
      {
        quartile: [add(1), add(-1)],
      }
    );
    expect(result).toEqual(expected);
    expectTypeOf(result).toEqualTypeOf<typeof expected>();
  });
```
In fact, when this test is run, it fails.

**Comment by fneco** [comment]
Thanks for your thoughtful review. The code has significantly improved as a result. I've addressed the noted issues, but there's one point I'd like to discuss further. Please see [my comment here (3)](https://github.com/remeda/remeda/pull/454#discussion_r1490324767) regarding that.

**Review by eranhirsch — APPROVED** [review]
Hey, thanks for your careful attention to detail and responsiveness with my feedback. I think this function turned out great and it's not easy to get the typing right for all edges cases this way.

I committed some minor nits to the typing based on my feedback and your responses so I can ship this without another review iteration.

I largely review PRs in Remeda alone, and have limited bandwidth for it so I can't always follow a complete review cycle as Remeda PRs usually require playing around with types quite a bit.

Thank you again for your contribution and I hope you find value in the library and continue to contribute when you find bugs or missing features.
